AsyncDisplayKit/Details/ASDataController.mm

```diff
@@ -535,16 +548,31 @@ - (void)waitUntilAllUpdatesAreCommitted
   return contexts;
 }
 
-- (void)invalidateDataSourceItemCounts
+- (void)invalidateDataSourceData
 {
   ASDisplayNodeAssertMainThread();
-  _itemCountsFromDataSourceAreValid = NO;
+  _dataFromDataSourceIsValid = NO;
 }
 
-- (std::vector<NSInteger>)itemCountsFromDataSource
+- (ASCellNodeBlock)nodeBlockForItemAtIndexPath:(NSIndexPath *)indexPath
 {
-  ASDisplayNodeAssertMainThread();
-  if (NO == _itemCountsFromDataSourceAreValid) {
+  if (self.supportsDeclarativeData) {
+    return [self.currentData elementOfKind:ASDataControllerRowNodeKind atIndexPath:indexPath].nodeBlock;
+  } else {
+    return [_dataSource dataController:self nodeBlockAtIndexPath:indexPath];
+  }
+}
+
+- (void)_fetchDataFromSourceIfNeeded
+{
+  if (_dataFromDataSourceIsValid) {
+    return;
+  }
+
+  if (self.supportsDeclarativeData) {
+    _data = [self.dataSource dataForDataController:self];
+    [_data markCompleted];
```

Comment:
We should update the layout spec API to be the same. Currently, setting `isMutable` to `NO`.